### ISSUE TITLE
[service] windows sdk

### ISSUE THREAD

**Issue by ericLemanissier** [conan-io/conan-center-index#4450]
@jgsogo @danimtb 
what version of the windows SDK are available on C3I workers building for Visual Studio 2019 (16.4.4+29728.190) ? (it would be interesting to get the SDK for other Visual Studio configurations too)

I'm asking this because https://github.com/conan-io/conan-center-index/pull/4412#issuecomment-769973980 has a surprising failure on Visual Studio 16 : `Windows SDK 10.0.18362.0 : 'C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um' not found or was incomplete` whereas this sdk is the one installed with VS 16 according to https://en.wikipedia.org/wiki/Microsoft_Windows_SDK
If I don't force the SDK version, `10.0.17763.0` is used, which seems incorrect, because it is the one provided with Visual Studio 2017.

**Comment by jgsogo** [comment]
Hi, @ericLemanissier 

I'm creating another job in the CI to check this kind of things, but as I'm not a Windows dev and I'm not running in Windows, it is a bit hard the try-fix-try loop. Maybe you (or someone) knows and can help.

I'm testing in the **staging environment** (production might have other versions), but I need first to know how to get the information.

If I list that directory I get different version (I assume those belong to different updates):

```
 Directory of C:\Program Files (x86)\Windows Kits\10\include

01/30/2020  07:18 PM    <DIR>          .
01/30/2020  07:18 PM    <DIR>          ..
03/25/2019  06:44 PM    <DIR>          10.0.10150.0
03/25/2019  06:56 PM    <DIR>          10.0.10240.0
03/25/2019  06:53 PM    <DIR>          10.0.10586.0
03/25/2019  08:24 PM    <DIR>          10.0.14393.0
01/30/2020  07:19 PM    <DIR>          10.0.17134.0
01/30/2020  07:17 PM    <DIR>          10.0.17763.0
```

but so far I need to know which one is the active one. 

I've found that there is a registry key:

```
REG QUERY "HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\Windows" 

HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\Windows
    CurrentVersion    REG_SZ    8.0.50727
    CurrentInstallFolder    REG_SZ    C:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\

HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v10.0
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v8.0
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v8.0A
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v8.1
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Microsoft SDKs\Windows\v8.1A
```

according to it, the default value is `8.0.50727`... 🤔   Is this default system-wide? Can each version of Visual Studio use a different default?

**Comment by ericLemanissier** [comment]
it looks like the sdk used depends on the build system. it seems cmake uses the most recent one by default, which led to my surprise that `10.0.18362` is not available while it should have been installed with Visual Studio 2019.

**Comment by jgsogo** [comment]
Maybe VS 2019 was installed before that Win SDK was available... and then it's been only updated.

In the new job, I can check that the Windows machines have some Win SDK available. Do you think if it is enough to check for the latest one? Right now I can hardcode `10.0.18362.0`, and when needed I can start to check for `10.0.19041.0`. I really don't know, but I would expect that the newer SDK _include_ the older ones, am I wrong?

**Comment by ericLemanissier** [comment]
I'm sorry, I don't know enough about windows sdks :/ I just know that there seem to be an "incompatibility" between `10.0.17763.0` and Visual Studio 2019. I don't know if there is some kind of forward or backward (or something else?) compatibility between different versions of the sdks.

**Comment by SSE4** [comment]
```
@REM Due to the SDK installer changes beginning with the 10.0.15063.0 (RS2 SDK), there is a chance that the
@REM Windows SDK installed may not have the full set of bits required for all application scenarios.
@REM We check for the existence of a file we know to be included in the "App" and "Desktop" portions
@REM of the Windows SDK, depending on the Developer Command Prompt's -app_platform configuration.
@REM If "windows.h" (UWP) or "winsdkver.h" (Desktop) are not found, the directory will be skipped as
@REM a candidate default value for [WindowsSdkDir].
set __check_file=winsdkver.h
if /I "%VSCMD_ARG_APP_PLAT%"=="UWP" set __check_file=Windows.h

if not "%WindowsSdkDir%"=="" for /f %%i IN ('dir "%WindowsSdkDir%include\" /b /ad-h /on') DO (
    @REM Skip if Windows.h|winsdkver (based upon -app_platform configuration) is not found in %%i\um.  
    if EXIST "%WindowsSdkDir%include\%%i\um\%__check_file%" (
        set result=%%i
        if "!result:~0,3!"=="10." (
            set SDK=!result!
            if "!result!"=="%VSCMD_ARG_WINSDK%" set findSDK=1
        )
    )
)
```

**Comment by SSE4** [comment]
seems like there are no traces of Windows SDK `10.0.18362` on the CI, neither in registry, nor in file system.
perhaps, it's coming from the Qt itself?
https://github.com/qt/qt5/blob/420d7f809c2b133456b24d047e54e29936d0180b/coin/pre-provisioning/qtci-windows-10-x86_64/msvc2019.txt#L69
https://github.com/qt/qt5/blob/ab23bc2ecf45c6003fdfac9c96cd56c268170648/coin/pre-provisioning/qtci-windows-10-x86/README_MANUAL_INSTALLATIONS#L45
https://github.com/qt/qt5/blob/ab23bc2ecf45c6003fdfac9c96cd56c268170648/coin/pre-provisioning/qtci-windows-10-x86/msvc2019.txt#L66
https://github.com/qt/qtwebengine/commit/723a0d074ef96083eecae58d3dbed78339a68402

**Comment by ericLemanissier** [comment]
> ```
> @REM Due to the SDK installer changes beginning with the 10.0.15063.0 (RS2 SDK), there is a chance that the
> @REM Windows SDK installed may not have the full set of bits required for all application scenarios.
> @REM We check for the existence of a file we know to be included in the "App" and "Desktop" portions
> @REM of the Windows SDK, depending on the Developer Command Prompt's -app_platform configuration.
> @REM If "windows.h" (UWP) or "winsdkver.h" (Desktop) are not found, the directory will be skipped as
> @REM a candidate default value for [WindowsSdkDir].
> set __check_file=winsdkver.h
> if /I "%VSCMD_ARG_APP_PLAT%"=="UWP" set __check_file=Windows.h
> 
> if not "%WindowsSdkDir%"=="" for /f %%i IN ('dir "%WindowsSdkDir%include\" /b /ad-h /on') DO (
>     @REM Skip if Windows.h|winsdkver (based upon -app_platform configuration) is not found in %%i\um.  
>     if EXIST "%WindowsSdkDir%include\%%i\um\%__check_file%" (
>         set result=%%i
>         if "!result:~0,3!"=="10." (
>             set SDK=!result!
>             if "!result!"=="%VSCMD_ARG_WINSDK%" set findSDK=1
>         )
>     )
> )
> ```

`/on` means it processes folders in alphabetical order, and the loop keeps the last item found, so it means that cmake (is this cmake code?) actually uses the latest sdk



> seems like there are no traces of Windows SDK `10.0.18362` on the CI, neither in registry, nor in file system.
> perhaps, it's coming from the Qt itself?

`Windows SDK 10.0.18362.0 : 'C:\Program Files (x86)\Windows Kits\10\include\10.0.18362.0\um' not found or was incomplete` happens only when [I specify the sdk version in the recipe ](https://github.com/conan-io/conan-center-index/commit/8f8875c824fee5117c2bfd915b118c17536861f3#diff-c548829da9d38ae15a8cb4447140bc50ac3f38250d1f348decd9569dc3298b24R470)
My surprise is that this sdk is not installed on C3I worker, when [it should be installed with Visual Studio 2019](https://en.wikipedia.org/wiki/Microsoft_Windows_SDK)
If I don't specify the sdk in the recipe, I get the error `C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\winrt\wrl\module.h(1427): error C2039: 'CheckForDuplicateEntries': is not a member of 'Microsoft::WRL::Details'` which happens because the 10.0.17763.0 sdk has some incompatibilities with visual studio 2019 and `/permissive-` flag

**Comment by SSE4** [comment]
> is this cmake code?
no, it's from vcvars implementation

> it should be installed with Visual Studio 2019
it says it's included, it doesn't necessarily mean it's installed by default (it needs to be checked)

**Comment by jgsogo** [comment]
There is always some default Win SDK that is installed together with Visual Studio, they are rolling that default version... and then you can optionally install more version. In the beginning, the newer version is optional and, once components start to depend on this new WinSDK version it is the default (you cannot opt-out it) and the older ones are optional (sorry I don't have the link, but it was some official MSFT response).

I'm not sure when VS 2019 was installed in these machines, maybe it was not the default by that time and the people who platform the machines didn't check those options. Right now it is `10.0.17763.0` the one installed. I've created a job to verify it.

I'll try to find who can update it (and modify the value in the job that checks). At least now we know the versions installed and we will be validating it.

**Comment by jgsogo** [comment]
After https://github.com/conan-io/conan-center-index/pull/4490 we have a job to run some checks against the nodes in the infrastructure. We have a version number hardcoded in the library/checks and we will check it is available in all the nodes:

```
OK  : Windows | windows1 | Win SDK 10.0.17763
OK  : Windows | windows2 | Win SDK 10.0.17763
OK  : Windows | windows3 | Win SDK 10.0.17763
```

We think (not 100% sure) that the latest number `.0` is something we don't need to check. I've seen some WinSDK version ending with `.1` and `.0`.

--- 

The next step will be to run this validation and check that the corresponding page in the documentation (https://github.com/conan-io/conan-center-index/blob/master/docs/supported_platforms_and_configurations.md) is synced.

**Comment by jgsogo** [comment]
Now that we have this check, and we understand what is happening here, does Qt6 needs a newer version of Win SDK? ... and, will it affect (break) other recipes? Is this the `glibc` for Windows 😅 ?

**Comment by dmn-star** [comment]
You could always use the latest Windows SDK version.

1)  Update Visual Studio
2)  Install the latest SDK from VS installer

> will it affect (break) other recipes?
No, it should not.

**Comment by ericLemanissier** [comment]
Its not really specific to qt, the compilation error happens when all of these are true:
- the project uses `permissive-` flag
- the project compiles under visual studio 2019
- the sdk used is `10.0.17763` or earlier
-  the project includes `winrt\wrl\module.h`

**Comment by SSE4** [comment]
> will it affect (break) other recipes?

very unlikely, it should not.

in general, Windows SDK contains the following:
- Windows API headers (e.g. windows.h, winsock.h, etc.)
- import libraries (no actual libraries, only stubs)
- some tools (rc - resource compiler, mt - manifest tool, signtool, debugging tools, etc.)

in theory, Microsoft could break something in headers, but it will be a massive disaster not just for us :)

I guess it's safe to roll Windows SDK periodically on our build slaves.

**Comment by SSE4** [comment]
> is this the `glibc` for Windows 😅 ?

not exactly. with Windows SDK, you may choose which Windows version to target: https://docs.microsoft.com/en-us/windows/win32/winprog/using-the-windows-headers. we currently don't model that in conan (we could, by adding `os.version`, but so far, no-one requested it).

**Comment by theirix** [comment]
Found another inconsistency with versions of Windows SDK and Visual Studio. My build #11262 is launched for `compiler.version=16` (it is a VS 2019). According to logs it uses Windows SDK 10.0.17763. Wiki says this SDK is for VS 2017. Seems like the compiler is very sensitive to Windows SDK version and the build breaks due to [new preprocessor implementation](https://docs.microsoft.com/en-us/cpp/preprocessor/preprocessor-experimental-overview?view=msvc-170) in MSVC 2019 and `/Zc:preprocessor`. Can the CCI builder choose the Windows SDK matching the Visual Studio version? 

Log can be seen [here](https://c3i.jfrog.io/c3i/misc/logs/pr/11262/4-configs/windows-visual_studio/libwebsockets/4.3.2//b683764b397174e19d98fc2750f2fe2b0dbb70f6-build.txt). Error excerpt:
```C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(487,17): warning C5103: pasting '/' and '/' does not result in a valid preprocessing token (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\shared\wtypes.h(745): message : in expansion of macro '_VARIANT_BOOL' (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(487,17): error C2059: syntax error: '/' (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(502,17): warning C5103: pasting '/' and '/' does not result in a valid preprocessing token (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\shared\wtypes.h(745): message : in expansion of macro '_VARIANT_BOOL' (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(502,17): error C2059: syntax error: '/' (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(529,40): warning C4094: untagged 'struct' declared no symbols (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(530,17): error C2059: syntax error: '}' (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(531,13): error C2059: syntax error: '}' (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(533,9): error C2059: syntax error: '}' (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(534,5): error C2059: syntax error: '}' (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(665,16): error C2079: 'varDefaultValue' uses undefined struct 'tagVARIANT' (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\oaidl.h(950,16): error C2079: 'varValue' uses undefined struct 'tagVARIANT' (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
C:\Program Files (x86)\Windows Kits\10\Include\10.0.17763.0\um\propidlbase.h(319,24): warning C5103: pasting '/' and '/' does not result in a valid preprocessing token (compiling source file C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\win32port\win32helpers\gettimeofday.c) [C:\J\w\prod\BuildSingleReference\.conan\data\libwebsockets\4.3.2\_\_\build\b683764b397174e19d98fc2750f2fe2b0dbb70f6\source_subfolder\lib\websockets_shared.vcxproj]
```

**Comment by jgsogo** [comment]
> Can the CCI builder choose the Windows SDK matching the Visual Studio version?

Hi! In the CI, the same Windows servers are used for all the VS versions, and AFAIL WindowsSDK is configured (defaulted) at the system level. Probably it is not a good idea to change this default for each one of the builds as there are several running in parallel using different Visual Studio versions.

I would suggest a different approach: if this library requires some specific WinSDK, probably this is something that can be configured in the build-system files (IIRC, it was possible in the `.vcproj` files, so probably it is possible in the CMake files as well). What do you think? Can you _explore_ this approach? Thanks!

**Comment by jgsogo** [comment]
@SSE4 dug a little bit more into this issue. It looks like there is no such thing as a _system default_, but there is some logic implemented in CMake to select one of the available WinSDK depending on VS version. Said that, if some specific version is needed, it has to be _forced_ in the build-system files (https://cmake.org/cmake/help/latest/variable/CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION_MAXIMUM.html).

**Comment by theirix** [comment]
I tried to explicitly select a newer Windows SDK for the machine which had two SDKs, and it [succeeds](https://c3i.jfrog.io/c3i/misc/logs/pr/11262/12-configs/windows-visual_studio/libwebsockets/4.3.2//98bea74c07190a7f5e639997d53e01ad804e8399-build.txt). It was done by setting a variable `CMAKE_SYSTEM_VERSION`. Unfortunately, it can't depict a range or a minimal version so it cannot be used in recipes. And CMake doesn't provide any information about which Windows SDK it had seen to select one.

Have no clues what happens on CCI machine and reproduce it. I tried to install two SDKs side by side on machine with VS 2019 and CMake chose a newest version as expected.
